### PR TITLE
Feature: Mineshaft Pity Display + OreMinedEvent

### PR DESCRIPTION
**Pull request by ItsEmpa**
## What
Adds a counter of your current pity counter for spawning a mineshaft. Also adds a OreMinedEvent that gets called when you mine a block using hypixel's custom mining system.

<details>
<summary>Images</summary>

image
image

</details>

## Changelog New Features
+ Added Mineshaft Pity Display. - Empa
    * Displays information about Mineshaft spawns, your progress towards a Pity Mineshaft, and more.

## Changelog Technical Details
+ Added OreBlock and OreType. - Empa
+ Added OreMinedEvent. - Empa
    * Provides information on the original block mined, and any other blocks mined by sources like Efficient Miner, Mole, etc.
+ Improved performance with Commission Color Blocks. - Empa

### CODE REVIEW

File: src/main/java/at/hannibal2/skyhanni/config/features/dev/DebugConfig.java [hunk 1]
```diff
@@ -126,6 +126,11 @@ public class DebugConfig {
     @ConfigEditorBoolean
     public boolean currentAreaDebug = true;
 
+    @Expose
+    @ConfigOption(name = "Oreblock Name", desc = "Show the OreBlock you are currently looking at")
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/config/features/dev/DebugConfig.java]:
```suggestion
    @ConfigOption(name = "Oreblock Name", desc = "Show the OreBlock you are currently looking at.")
```

File: src/main/java/at/hannibal2/skyhanni/config/features/mining/MineshaftPityDisplay.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        MineshaftPityLines.TIME_SINCE_MINESHAFT
+    ));
+
+    @Expose
+    @ConfigOption(name = "Modify Spawn Message", desc = "Modify the Mineshaft spawn message with more stats.")
+    @ConfigEditorBoolean
+    @FeatureToggle
+    public boolean modifyChatMessage = true;
+
+
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/config/features/mining/MineshaftPityDisplay.java]:
```suggestion

```

File: src/main/java/at/hannibal2/skyhanni/config/features/mining/MineshaftPityDisplay.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class MineshaftPityDisplay {
+    @Expose
+    @ConfigOption(name = "Enable Display", desc = "Enable the Mineshaft Pity Display.")
+    @ConfigEditorBoolean
+    @FeatureToggle
+    public boolean enable = true;
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/config/features/mining/MineshaftPityDisplay.java]: This should be enabled

File: src/main/java/at/hannibal2/skyhanni/events/mining/CustomBlockMineEvent.kt [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package at.hannibal2.skyhanni.events.mining
+
+import at.hannibal2.skyhanni.events.LorenzEvent
+import at.hannibal2.skyhanni.features.mining.OreBlock
+
+class CustomBlockMineEvent(val originalOre: OreBlock, val map: Map<OreBlock, Int>) : LorenzEvent()
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/events/mining/CustomBlockMineEvent.kt]: the map isnt used and this function should probably be called OreMinedEvent

File: src/main/java/at/hannibal2/skyhanni/utils/LorenzVec.kt [hunk 5]
```diff
@@ -128,6 +128,8 @@ data class LorenzVec(
 
     fun toCleanString(): String = "$x $y $z"
 
+    fun toCleanStringWithSeparators(): String = "$x, $y, $z"
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/utils/LorenzVec.kt]: maybe this can have a separator parameter that is default set to a comma

File: src/main/java/at/hannibal2/skyhanni/data/MiningAPI.kt [hunk 6]
```diff
@@ -29,13 +40,37 @@ object MiningAPI {
         "§c ☠ §r§7§r§.(?<name>.+)§r§7 (?<reason>.+)"
     )
 
+    class MinedBlock(val ore: OreBlock, var position: LorenzVec, var confirmed: Boolean, val time: SimpleTimeMark)
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/data/MiningAPI.kt]: probably can be a data class

File: src/main/java/at/hannibal2/skyhanni/data/MiningAPI.kt [hunk 7]
```diff
@@ -29,13 +40,37 @@ object MiningAPI {
...
+
+    var isBeingMined = false
+    var lastSound = SimpleTimeMark.farPast()
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/data/MiningAPI.kt]: Can both be private

File: src/main/java/at/hannibal2/skyhanni/data/MiningAPI.kt [hunk 8]
```diff
@@ -29,13 +40,37 @@ object MiningAPI {
...
+
+    private var recentMinedBlocksMap = mutableListOf<MinedBlock>()
+    private var surroundingMinedBlocks = mutableListOf<MinedBlock>()
+    private val allowedSoundNames = listOf("dig.glass", "dig.stone", "dig.gravel", "dig.cloth")
+
     private var cold = 0
     var lastColdUpdate = SimpleTimeMark.farPast()
     var lastColdReset = SimpleTimeMark.farPast()
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/data/MiningAPI.kt:80]:
```suggestion
```

File: src/main/java/at/hannibal2/skyhanni/features/mining/MineshaftPityDisplay.kt [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+
+    enum class MineshaftPityLines(private val display: String, val shouldDisplay: () -> Boolean = { true }) {
+        TITLE("§3§lMineshaft Pity Counter"),
+        COUNTER("§3Counter: §e561§6/§e2000"),
+        CHANCE("§3Chance: §e1§6/§e1439 §7(§b0.069%§7)"),
+        NEEDED_TO_PITY("§3Needed to pity:\n§7   <blocks>"),
+        TIME_SINCE_MINESHAFT("§3Last Mineshaft: §e21m 5s", { !lastMineshaftSpawn.isFarPast() }),
+        AVERAGE_BLOCKS_MINESHAFT("§3Average Blocks/Mineshaft: §e361.5", { mineshaftTotalCount != 0 })
+
+        ;
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/mining/MineshaftPityDisplay.kt]:
```suggestion
        AVERAGE_BLOCKS_MINESHAFT("§3Average Blocks/Mineshaft: §e361.5", { mineshaftTotalCount != 0 }),
        ;
```

**Review by CalMWolfs — CHANGES_REQUESTED** [review]
a few code cleanup things

**Comment by hannibal002** [comment]
new pity changes needed

**Comment by ItsEmpa** [comment]
marking as draft since something seems to have broken, again

**Comment by ItsEmpa** [comment]
ready for review again, surely this is the one

**Comment by Nealkitt** [comment]
Another nitpick, i've done some testing and the pity seems to reset on lobby swap, i can easily get negative numbers by just lobby swapping a lot
image

File: src/main/java/at/hannibal2/skyhanni/config/features/dev/DebugConfig.java [hunk 10]
```diff
@@ -126,6 +126,11 @@ public class DebugConfig {
     @ConfigEditorBoolean
     public boolean currentAreaDebug = true;
 
+    @Expose
+    @ConfigOption(name = "Oreblock Name", desc = "Show the OreBlock you are currently looking at.")
+    @ConfigEditorBoolean
+    public boolean raytracedOreblock = true;
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/config/features/dev/DebugConfig.java:132]: is this dev feature meant to be default on?

**Review by CalMWolfs — CHANGES_REQUESTED** [review]
just this one thing

File: src/main/java/at/hannibal2/skyhanni/config/features/dev/DebugConfig.java (hunk 10 above)
Comment (ItsEmpa) [reply src/main/java/at/hannibal2/skyhanni/config/features/dev/DebugConfig.java:132]: this is only shown when in f3 mode, so i didnt really think much about it, but sure i can make it default off

**Review by CalMWolfs — APPROVED** [review]

File: src/main/java/at/hannibal2/skyhanni/data/MiningAPI.kt [hunk 11]
```diff
@@ -81,4 +267,38 @@ object MiningAPI {
...
+
+            inGlacite = newInGlacite
+            inDwarvenMines = newInDwarvenMines
+            inCrystalHollows = newInCrystalHollows
+            inCrimsonIsle = newInCrimsonIsle
+            inEnd = newInEnd
+            inSpidersDen = newInSpidersDen
+
+            val entries = OreBlock.entries
+            val oreBlockList: List<OreBlock> = if (newInGlacite) entries.filter { it.checkArea.invoke() }
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/data/MiningAPI.kt]: here are many things weird:
it looks like we update this every tick, instead of every time the user changes the area?
why are there non mining islands?
why is every case doing the same thing?
you can remove .invoke,
you can use switch instead of if else chain

**Review by hannibal002 — CHANGES_REQUESTED** [review]
one thing

**Review by hannibal002 — APPROVED** [review]
Works for me
